cherab/core/laser/node.pyx

```diff
@@ -0,0 +1,249 @@
+    from the Laser object.
+
+    The main physical properties of the laser are defined by the three
+    attributes laser_spectrum, laser_profile and models. The laser_spectrum
+    has to be an instance of LaserSpectrum and defines the spectral properties
+    of the laser light. The laser_profile has to be an instance of LaserProfile
+    and it holds all the space related definitions as volumetric distribution
+    of laser light energy polarisation direction. In the models a list of LaserModels
+    can be stored, which calculate the contribution of the laser ligth to the observed
+    radiation. The models can cover various applications as for example
+    Thomson scattering. Please see the documentation of individual classes
+    for more detail.
+
+    The shape of the laser (e.g. cylinder) and its parameters (e.g. radius)
+    is controled by the LaserProfile.
+
+    The plasma reference has to be specified to attach the any models.
+
+    :param Node parent: The parent node in the Raysect scene-graph.
+      See the Raysect documentation for more guidance.
+    :param AffineMatrix3D transform: The transform defining the spatial position
+      and orientation of this laser. See the Raysect documentation if you need
+      guidance on how to use AffineMatrix3D transforms.
+    :param str name: The name for this laser object.
+    :ivar Plasma plasma: The plasma instance with which this laser interacts.
+    :ivar float importance: The importance sampling factor.
+    :ivar LaserSpectrum laser_spectrum: The LaserSpectrum instance with which this laser interacts.
+    :ivar LaserProfile laser_profile: The LaserProfile instance with which this laser interacts.
+    :ivar ModelManager models: The manager class that sets and provides access to the
+      emission models for this laser.
+    :ivar VolumeIntegrator integrator: The configurable method for doing
+      volumetric integration through the laser along a Ray's path. Defaults to
+      a numerical integrator with 1mm step size, NumericalIntegrator(step=0.001).
+    """
+
+    def __init__(self, object parent=None, AffineMatrix3D transform=None, str name=None):
+
+        super().__init__(parent, transform, name)
+
+        self._set_init_values()
+
+        self.notifier = Notifier()
+
+        self._models = ModelManager()
+        
+        self._integrator = NumericalIntegrator(step=1e-3)
+
+        self._importance = 1.
+        self._configure()
+
+    def _set_init_values(self):
+        """
+        Sets initial values of the laser shape to avoid errors.
+        """
+        self._importance = 0.
+        self._geometry = []
+
+    @property
+    def plasma(self):
+        return self._plasma
+
+    @plasma.setter
+    def plasma(self, Plasma value not None):
+
+        #unregister from old plasma notifier
+        if self._plasma is not None:
+            self._plasma.notifier.remove(self._plasma_changed)
+
+        self._plasma = value
```

Comment:
Thanks for spotting this! This could definitely lead to serious problems. `Plasma` has to notify `Laser`, so it rebuilds the materials, because  `LaserMaterial` caches transforms to plasma.  I separated the method `_configure_geometry` into `_build_geometry`, which now only builds the primitives and into `_configure_materials`, which updates the materials. The `_configure_geometry` was replaced by `configure_geometry`